Add koji-builder CRD

Zlopez opened the pull request:
Custom Resource Definition for koji-builder component. Basic structure and tests.

Zlopez commented:
Local and cluster tests are working, except pod deployment, which is failing on connecting to running koji-hub

Zlopez commented:
So this needs to wait till #10 is finished

odra commented on mbox-operator/roles/koji-builder/defaults/main.yml:
```diff
@@ -0,0 +1,15 @@
+---
+# defaults file for koji-builder
+koji_builder_image: quay.io/fedora/koji-builder:latest
```
can you use a value from the cr the same as the others "image | default('foobar')"

odra replied:
the same applies for the other fields

odra commented on mbox-operator/roles/koji-builder/tasks/cert.yml:
```diff
@@ -0,0 +1,61 @@
+    - openssl_privatekey:
+        path: /tmp/{{ cert_name }}_client_key.pem
+        size: 4096
+    - openssl_csr:
```
`common_name`needs to be in here because koji-hub will use its value as username, it might also be a good idea to set this value from a cr property

Zlopez replied:
And the `common_name` needs to be same as in koji-hub?

odra replied:
yeah some username, you can use koji_hub, koji-hub will use it as the api requester username

odra commented:
Could you also update our `Makefile` to create and delete the builder crds?

odra commented on mbox-operator/roles/koji-builder/templates/configmap.yml.j2:
```diff
@@ -0,0 +1,23 @@
+
+    ; Configuration authentication
+    serverca = /etc/cacert/cert
+    cert=/etc/clientcert/tls.cert
```
should be `/etc/clientcert/tls.crt` ("crt" intead of "cert")

odra replied:
here we need to use a file that contains both key and cert (key being the first one) separated by a '\n' char.

Zlopez replied:
I was wondering what `combined` meant in the original config.

Zlopez replied:
Looking at the original `ca.py` code I'm now clearly see it. I'm wondering why I didn't saw this before :-(

Zlopez replied:
I tried to create a combined certificate using this:
```
tls_combined.pem: "{{ lookup('file', cert_dir.path + '/client_cert.pem') | b64encode }}\n{{ lookup('file', cert_dir.path + '/client_key.pem') | b64encode }}"
```
I hope this is correct approach.

odra replied:
you need  concatenate them all before encoding to base64:
```
"{{ (lookup('file', cert_dir.path + '/client_cert.pem') + '\n' + lookup('file', cert_dir.path + '/ca_cert.pem')) | b64encode }}"}"
```

odra commented on mbox-operator/roles/koji-builder/templates/configmap.yml.j2:
```diff
@@ -0,0 +1,23 @@
+    plugins=
+
+    ; Configuration authentication
+    serverca = /etc/cacert/cert
```
so we have to create a file that contains two certs: the root one + koji-builder one. (split by a '\n') instead of  using the  raw root cert.

it might be a good idea to create a new entry in the secret called "ca_chain.pem" so you can use it directly here.

Zlopez replied:
I will try it, but looking at the `ca.py` the cert is just plain ca.crt

Zlopez commented on mbox-operator/roles/koji-builder/tasks/cert.yml:
```diff
@@ -0,0 +1,61 @@
+- set_fact:
```
This was changed with latest commit.

Zlopez commented:
I added the combined cert for both `serverca` and `cert` field in configuration, but I'm still getting `SSLError`, I just don't know what is wrong.

Zlopez commented:
Ok, so I found that the way I created the combined cert was wrong and used the one created in `startup.sh` script. Now I'm getting this issue with authentication, but it seems that this is related to koji-hub configuration of Apache:
```
[Wed May 06 11:34:54.781103 2020] [ssl:info] [pid 26:tid 140469078042368] [client 172.18.0.9:57372] AH01964: Connection to child 138 established (server mbox.dev:443)
[Wed May 06 11:34:54.781352 2020] [ssl:debug] [pid 26:tid 140469078042368] ssl_engine_kernel.c(2391): [client 172.18.0.9:57372] AH02044: No matching SSL virtual host for servername koji-hub found (using default/first virtual host)
[Wed May 06 11:34:54.788700 2020] [ssl:debug] [pid 26:tid 140469078042368] ssl_engine_kernel.c(2254): [client 172.18.0.9:57372] AH02041: Protocol: TLSv1.2, Cipher: ECDHE-RSA-AES256-GCM-SHA384 (256/256 bits)
[Wed May 06 11:34:54.789689 2020] [ssl:info] [pid 26:tid 140469078042368] (70014)End of file found: [client 172.18.0.9:57372] AH01991: SSL input filter read failed.
[Wed May 06 11:34:54.789867 2020] [ssl:debug] [pid 26:tid 140469078042368] ssl_engine_io.c(1102): [client 172.18.0.9:57372] AH02001: Connection closed to child 138 with standard shutdown (server mbox.dev:443)
```

Zlopez commented:
The koji-builder PR should be ready for review and merge.

The deployment is still failing, because of the issue when connecting to koji-hub, but we decided to merge it anyway and solve the issue separately.

AdamSaleh commented on mbox-operator/roles/koji-builder/tasks/cert.yml:
```diff
@@ -0,0 +1,71 @@
+        namespace: "{{ meta.namespace }}"
+      register: k8s_secrets
+    - fail:
+          msg: "Secret {{ koji_builder_cacert_secret }} not found."
```
Wrong indent, according to yaml-lint.

Zlopez replied:
Fixed now

Zlopez commented on mbox-operator/deploy/crds/apps.fedoraproject.org_v1alpha1_mbkojihub_cr.yaml:
```diff
@@ -8,7 +8,7 @@ spec:
   image: quay.io/fedora/koji-hub:latest
   replicas: 1
   persistent: true
-  host: 'mbox.dev'
```
This needs to be set to same as in koji-hub, so I think this will be ideal candidate to move to shared mbox role. For now I will let it here.

Zlopez commented on mbox-operator/molecule/test-cluster/converge/converge_koji-builder.yml:
```diff
@@ -0,0 +1,143 @@
+    ansible_python_interpreter: '{{ ansible_playbook_python }}'
+    deploy_dir: "{{ lookup('env', 'MOLECULE_PROJECT_DIRECTORY') }}/deploy"
+    pull_policy: Never
+    REPLACE_IMAGE: apps.fedoraproject.org/mbox-operator:testing
```
I tried it only on my local minikube, where it worked.

Zlopez replied:
I see, that this is now causing failure on CI, how we could fix this?

Zlopez commented:
I fixed the tests for local run and fixed the yamllint issue.

To allow the test pass I commented out the deployment and assert that checks the deployed pod.

odra commented on mbox-operator/molecule/default/create/koji-builder.yml:
```diff
@@ -0,0 +1,21 @@
+        namespace: "{{ namespace }}"
+        definition: "{{ lookup('file', custom_resource) | from_yaml }}"
+        wait: true
+        wait_condition:
```
that part is what is making it fail right? since koji-builder is never ready.

Zlopez replied:
It fails with this on the cluster:
```
2020-05-12 10:53:01.961929 | cloud-host |     TASK [Get existing image hash] *************************************************
2020-05-12 10:53:01.962098 | cloud-host |     task path: /home/zuul-worker/src/github.com/fedora-infra/mbbox/mbox-operator/molecule/test-cluster/converge/converge_koji-builder.yml:11
2020-05-12 10:53:01.975151 | cloud-host |     <127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: zuul-worker
2020-05-12 10:53:01.975539 | cloud-host |     <127.0.0.1> EXEC /bin/sh -c 'echo ~zuul-worker && sleep 0'
2020-05-12 10:53:01.980862 | cloud-host |     <127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/zuul-worker/.ansible/tmp `"&& mkdir /home/zuul-worker/.ansible/tmp/ansible-tmp-1589280781.9804547-21495-141410314518775 && echo ansible-tmp-1589280781.9804547-21495-141410314518775="` echo /home/zuul-worker/.ansible/tmp/ansible-tmp-1589280781.9804547-21495-141410314518775 `" ) && sleep 0'
2020-05-12 10:53:02.091313 | cloud-host |     Using module file /usr/local/lib/python3.7/site-packages/ansible/modules/commands/command.py
2020-05-12 10:53:02.092339 | cloud-host |     <127.0.0.1> PUT /home/zuul-worker/.ansible/tmp/ansible-local-21386je5ilzeu/tmpvhj4p5pd TO /home/zuul-worker/.ansible/tmp/ansible-tmp-1589280781.9804547-21495-141410314518775/AnsiballZ_command.py
2020-05-12 10:53:02.093201 | cloud-host |     <127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/zuul-worker/.ansible/tmp/ansible-tmp-1589280781.9804547-21495-141410314518775/ /home/zuul-worker/.ansible/tmp/ansible-tmp-1589280781.9804547-21495-141410314518775/AnsiballZ_command.py && sleep 0'
2020-05-12 10:53:02.099055 | cloud-host |     <127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 /home/zuul-worker/.ansible/tmp/ansible-tmp-1589280781.9804547-21495-141410314518775/AnsiballZ_command.py && sleep 0'
2020-05-12 10:53:02.283858 | cloud-host |     <127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/zuul-worker/.ansible/tmp/ansible-tmp-1589280781.9804547-21495-141410314518775/ > /dev/null 2>&1 && sleep 0'
2020-05-12 10:53:02.292701 | cloud-host | fatal: [localhost]: FAILED! => {
2020-05-12 10:53:02.292728 | cloud-host |     "changed": false,
2020-05-12 10:53:02.292733 | cloud-host |     "cmd": [
2020-05-12 10:53:02.292738 | cloud-host |         "docker",
2020-05-12 10:53:02.292756 | cloud-host |         "images",
2020-05-12 10:53:02.292764 | cloud-host |         "-q",
2020-05-12 10:53:02.293495 | cloud-host |         "apps.fedoraproject.org/mbox-operator:testing"
2020-05-12 10:53:02.293510 | cloud-host |     ],
2020-05-12 10:53:02.293515 | cloud-host |     "delta": "0:00:00.054787",
2020-05-12 10:53:02.293529 | cloud-host |     "end": "2020-05-12 10:53:02.273513",
2020-05-12 10:53:02.293534 | cloud-host |     "invocation": {
2020-05-12 10:53:02.293539 | cloud-host |         "module_args": {
2020-05-12 10:53:02.293543 | cloud-host |             "_raw_params": "docker images -q apps.fedoraproject.org/mbox-operator:testing",
2020-05-12 10:53:02.293548 | cloud-host |             "_uses_shell": false,
2020-05-12 10:53:02.293552 | cloud-host |             "argv": null,
2020-05-12 10:53:02.293556 | cloud-host |             "chdir": null,
2020-05-12 10:53:02.293561 | cloud-host |             "creates": null,
2020-05-12 10:53:02.293565 | cloud-host |             "executable": null,
2020-05-12 10:53:02.293569 | cloud-host |             "removes": null,
2020-05-12 10:53:02.293574 | cloud-host |             "stdin": null,
2020-05-12 10:53:02.293578 | cloud-host |             "stdin_add_newline": true,
2020-05-12 10:53:02.293582 | cloud-host |             "strip_empty_ends": true,
2020-05-12 10:53:02.293587 | cloud-host |             "warn": true
2020-05-12 10:53:02.293591 | cloud-host |         }
2020-05-12 10:53:02.293595 | cloud-host |     },
2020-05-12 10:53:02.293600 | cloud-host |     "msg": "non-zero return code",
2020-05-12 10:53:02.293604 | cloud-host |     "rc": 1,
2020-05-12 10:53:02.293608 | cloud-host |     "start": "2020-05-12 10:53:02.218726",
2020-05-12 10:53:02.293613 | cloud-host |     "stderr": "Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?",
2020-05-12 10:53:02.293617 | cloud-host |     "stderr_lines": [
2020-05-12 10:53:02.293621 | cloud-host |         "Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?"
2020-05-12 10:53:02.293626 | cloud-host |     ],
2020-05-12 10:53:02.293631 | cloud-host |     "stdout": "",
2020-05-12 10:53:02.293635 | cloud-host |     "stdout_lines": []
2020-05-12 10:53:02.293639 | cloud-host | }
2020-05-12 10:53:02.295105 | cloud-host |
2020-05-12 10:53:02.295123 | cloud-host |     PLAY RECAP *********************************************************************
2020-05-12 10:53:02.295289 | cloud-host |     localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

odra replied:
it worked for me locally, let me try in a cluster

odra replied:
the error says it can't connect to the docker daemon, did it die due to lack of resources?

Zlopez replied:
I tried the cluster tests on local minikube and it's failing miserably. 
* CR for koji-hub can't be created, because postgres secret doesn't exist.
* CR for koji-builder can't be created, because ca cert that should be created by koji-hub is not in place
* image for mbox-operator can't be build, this is caused not finding the path, which is probably related to what @AdamSaleh said above

odra replied:
I thought we were running local test only for now, shouldn't that stuff be addressed in the test re-factoring issue/prs?

odra replied:
there are no cluster tests right now  for koji-hub as well and I think it will require some extra work to make it happen which is why I think it should be addressed on other PRs.

AdamSaleh commented:
New plan, as I am trying to continue on top of this anyway, what if we just disable the CI for now (the test-cluster job, docs are fine) and once I fix this I enable the test-cluster in my PR again.

Zlopez commented:
This should be fine, because you are working on the cluster test anyway. At least this wouldn't block you anymore.

Zlopez commented:
I commented out the `create` and `converge` sequence from `test-cluster`.
